### PR TITLE
Separate IResult based results from ActionResults

### LINKED ISSUE #33729: Create IResult-returning static methods
`IResult`-returning static methods would allow something like `using static Microsoft.AspNetCore.Http.Results` and makes for better return type inference for IResult-returning lambdas, so you don't have to do casting like the following you have to do using constructors today.

```c#
app.MapGet("/todos/{id}", async (int id) =>
{
    using var db = new TodoDbContext(options);
 
    if (await db.Todos.FindAsync(id) is not Todo todo)
    {
        return (IResult)new NotFoundResult();
    }
 
    return new JsonResult(todo);
})
```

With, `IResult` returning `NotFound` and `Ok` methods the `(IResult)` casting can be removed.

```c#
app.MapGet("/todos/{id}", async (int id) =>
{
    using var db = new TodoDbContext(options);
 
    if (await db.Todos.FindAsync(id) is not Todo todo)
    {
        return NotFound();
    }
 
    return Json(todo);
})
```

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #33729

### CODE REVIEW

**Comment by davidfowl** [comment]
Why a new assembly?

**Comment by pranavkm** [comment]
The results depend on M.A.Routing which adds a circular dependency if they are in M.A.H.Extensions.

File: src/Http/Http.Results/src/ObjectResult.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        public object? Value { get; }
+
+        /// <summary>
+        /// Gets or sets the HTTP status code.
+        /// </summary>
+        public int? StatusCode { get; }
+
+        public Task ExecuteAsync(HttpContext httpContext)
+        {
+            var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
```
Comment (davidfowl) [inline src/Http/Http.Results/src/ObjectResult.cs:33]: We're going to want to cache these calls to the logger, but that should be in another PR. My guess is that we haven't looked at this cost in MVC.

File: src/Middleware/StaticFiles/src/LoggerExtensions.cs [hunk 2]
```diff
@@ -12,19 +12,19 @@ namespace Microsoft.AspNetCore.StaticFiles
...
-        private static Action<ILogger, string, Exception?> _fileNotFound;
-        private static Action<ILogger, string, Exception?> _fileNotModified;
-        private static Action<ILogger, string, Exception?> _preconditionFailed;
-        private static Action<ILogger, int, string, Exception?> _handled;
-        private static Action<ILogger, string, Exception?> _rangeNotSatisfiable;
-        private static Action<ILogger, StringValues, string, Exception?> _sendingFileRange;
-        private static Action<ILogger, StringValues, string, Exception?> _copyingFileRange;
-        private static Action<ILogger, Exception?> _writeCancelled;
-        private static Action<ILogger, Exception?> _endpointMatched;
+        private static readonly Action<ILogger, string, Exception?> _methodNotSupported;
```
Comment (davidfowl) [inline src/Middleware/StaticFiles/src/LoggerExtensions.cs:15]: Source generator?

File: src/Http/Http.Results/src/Results.cs [hunk 3]
```diff
@@ -0,0 +1,1243 @@
...
+        /// Creates a <see cref="AcceptedAtRouteResult"/> object that produces an <see cref="StatusCodes.Status202Accepted"/> response.
+        /// </summary>
+        /// <param name="routeValues">The route data to use for generating the URL.</param>
+        /// <param name="value">The optional content value to format in the entity body; may be null.</param>
+        /// <returns>The created <see cref="AcceptedAtRouteResult"/> for the response.</returns>
+        public static IResult AcceptedAtRoute(object? routeValues, object? value)
+            => AcceptedAtRoute(routeName: null, routeValues: routeValues, value: value);
+
+        /// <summary>
+        /// Creates a <see cref="AcceptedAtRouteResult"/> object that produces an <see cref="StatusCodes.Status202Accepted"/> response.
```
Comment (DamianEdwards) [inline src/Http/Http.Results/src/Results.cs:1233]: Need to swap your 'a' and 'an' around, e.g. ` Creates an <see cref="AcceptedAtRouteResult"/> object that produces a <see cref="StatusCodes.Status202Accepted"/> response`

File: src/Http/Http.Results/src/Results.cs [hunk 4]
```diff
@@ -0,0 +1,1243 @@
...
+            }
+
+            return new AcceptedResult(locationUri: uri, value: null);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="AcceptedResult"/> object that produces an <see cref="StatusCodes.Status202Accepted"/> response.
+        /// </summary>
+        /// <param name="uri">The optional URI with the location at which the status of requested content can be monitored.
+        /// May be null.</param>
```
Comment (DamianEdwards) [inline src/Http/Http.Results/src/Results.cs:1168]: We're not consistent with how we document nullable ref types in params. Some are 'blah blah; may be null', others are like this one. Let's just pick one form and stick with it.

File: src/Http/Http.Results/src/Results.cs [hunk 5]
```diff
@@ -0,0 +1,1243 @@
...
+        /// Creates a <see cref="AcceptedResult"/> object that produces an <see cref="StatusCodes.Status202Accepted"/> response.
+        /// </summary>
+        /// <returns>The created <see cref="AcceptedResult"/> for the response.</returns>
+        public static IResult Accepted()
+            => new AcceptedResult();
+
+        /// <summary>
+        /// Creates a <see cref="AcceptedResult"/> object that produces an <see cref="StatusCodes.Status202Accepted"/> response.
+        /// </summary>
+        /// <param name="value">The optional content value to format in the entity body; may be null.</param>
```
Comment (DamianEdwards) [inline src/Http/Http.Results/src/Results.cs:1143]: Might 'response body' be clearer language than 'entity body'?

File: src/Http/Http.Results/src/Results.cs [hunk 6]
```diff
@@ -0,0 +1,1243 @@
...
+        /// <param name="contentType">The Content-Type of the file.</param>
+        /// <param name="fileDownloadName">The suggested file name.</param>
+        /// <param name="enableRangeProcessing">Set to <c>true</c> to enable range requests processing.</param>
+        /// <returns>The created <see cref="PhysicalFileResult"/> for the response.</returns>
+        public static IResult PhysicalFile(
+            string physicalPath,
+            string contentType,
+            string? fileDownloadName,
+            bool enableRangeProcessing)
+            => new PhysicalFileResult(physicalPath, contentType)
```
Comment (DamianEdwards) [inline src/Http/Http.Results/src/Results.cs:590]: Curious, why is this one expression bodied but the next one not?

File: src/Http/Http.Results/src/Results.cs [hunk 7]
```diff
@@ -0,0 +1,1243 @@
...
+    ///
+    /// </summary>
+    public static class Results
+    {
+        #region Challenge
+
+        /// <summary>
+        /// Creates a <see cref="ChallengeResult"/>.
+        /// </summary>
+        /// <returns>The created <see cref="ChallengeResult"/> for the response.</returns>
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Http/Http.Results/src/Results.cs:24]: The API docs will referer to the type full name, and completion colorizes it correctly, but you're not able to control click on the type to get there. Not super weird. I didn't change the doc comments much as I mas transferring them from MVC, but if you feel strongly about it, I can do a pass to clean it up.

File: src/Http/Http.Results/src/Results.cs (hunk 6 above)
Comment (pranavkm) [reply src/Http/Http.Results/src/Results.cs:590]: Somebody forgot to update MVC: https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/ControllerBase.cs#L1647-L1677

File: src/Http/Http.Results/src/Results.cs (hunk 5 above)
Comment (pranavkm) [reply src/Http/Http.Results/src/Results.cs:1143]: Sure. I think this comment might have been from WebAPI2 which tends to use the [RFC names](https://www.w3.org/Protocols/rfc2616/rfc2616-sec7.html)

File: src/Middleware/StaticFiles/src/LoggerExtensions.cs (hunk 2 above)
Comment (pranavkm) [reply src/Middleware/StaticFiles/src/LoggerExtensions.cs:15]: It's this analyzer - which is a suggestion, but shows up as warnings in VS: https://github.com/dotnet/aspnetcore/blob/main/.editorconfig#L100-L101. Source generator to happen separately.

**Comment by davidfowl** [comment]
@pranavkm can you make one of the samples use these?

**Comment by dougbu** [comment]
Curious: I skimmed the projects but couldn't find a reference to the new project. Other than the samples @davidfowl suggested, where will the assembly be used❔

File: src/Http/Http.Results/src/PhyiscalFileResult.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Http.Result
+{
+    /// <summary>
+    /// A <see cref="PhysicalFileResult"/> on execution will write a file from disk to the response
+    /// using mechanisms provided by the host.
+    /// </summary>
+    internal sealed partial class PhysicalFileResult : FileResult, IResult
```
Comment (martincostello) [inline src/Http/Http.Results/src/PhyiscalFileResult.cs:17]: Not quite the right line for this comment, but the file name has a typo: `PhyiscalFileResult.cs` => `PhysicalFileResult.cs`

**Comment by pranavkm** [comment]
@dougbu the plan is to use these in the minimal actions project template. I updated a functional test to use some of these result types.

File: src/Http/Http.Results/src/ChallengeResult.cs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            public static void ChallengeResultExecuting(ILogger logger, IList<string> authenticationSchemes)
+            {
+                if (logger.IsEnabled(LogLevel.Information))
+                {
+                    ChallengeResultExecuting(logger, authenticationSchemes.ToArray());
+                }
+            }
+
+            [LoggerMessage(1, LogLevel.Information, "Executing ChallengeResult with authentication schemes ({Schemes}).", EventName = "ChallengeResultExecuting")]
```
Comment (martincostello) [inline src/Http/Http.Results/src/ChallengeResult.cs:116]: Can use `SkipEnabledCheck` here.

File: src/Http/Http.Results/src/PhysicalFileResult.cs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        // For testing
+        public Func<string, FileInfoWrapper> GetFileInfoWrapper { get; init; } =
+            static path => new FileInfoWrapper(path);
+
+        public Task ExecuteAsync(HttpContext httpContext)
+        {
+            var fileInfo = GetFileInfoWrapper(FileName);
+            if (!fileInfo.Exists)
+            {
+                throw new FileNotFoundException($"Could not find file: {FileName}");
```
Comment (martincostello) [inline src/Http/Http.Results/src/PhysicalFileResult.cs:45]:
```suggestion
                throw new FileNotFoundException($"Could not find file: {FileName}", FileName);
```

File: src/Http/Http.Results/src/AcceptedAtRouteResult.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            object? routeValues,
+            object? value)
+            : base(value, StatusCodes.Status202Accepted)
+        {
+            RouteName = routeName;
+            RouteValues = new RouteValueDictionary(routeValues);
+        }
+
+        /// <summary>
+        /// Gets or sets the name of the route to use for generating the URL.
```
Comment (martincostello) [inline src/Http/Http.Results/src/AcceptedAtRouteResult.cs:41]:
```suggestion
        /// Gets the name of the route to use for generating the URL.
```

File: src/Http/Http.Results/src/AcceptedAtRouteResult.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        /// </summary>
+        public string? RouteName { get; }
+
+        /// <summary>
+        /// Gets or sets the route data to use for generating the URL.
```
Comment (martincostello) [inline src/Http/Http.Results/src/AcceptedAtRouteResult.cs:46]:
```suggestion
        /// Gets the route data to use for generating the URL.
```

File: src/Http/Http.Results/src/ChallengeResult.cs (hunk 9 above)
Comment (pranavkm) [reply src/Http/Http.Results/src/ChallengeResult.cs:116]:
```suggestion
            [LoggerMessage(1, LogLevel.Information, "Executing ChallengeResult with authentication schemes ({Schemes}).", EventName = "ChallengeResultExecuting", SkipEnabledCheck = true)]
```

**Review by halter73 — APPROVED** [review]
